tests/core/tests/test_resources.py

```diff
@@ -2570,3 +2571,143 @@ def test_has_natural_foreign_key(self):
+        book_resource._meta.skip_unchanged = True
+
+        result = book_resource.import_data(dataset, dry_run=False)
+
+        self.assertFalse(result.has_errors())
+        self.assertEqual(len(result.rows), 1)
+
+    def test_m2m_on_existing_entity_adding_relations(self):
+        class _BookM2MResource(resources.ModelResource):
+            categories = fields.Field(
+                attribute='categories',
+                m2m_add=True,
+                widget=widgets.ManyToManyWidget(Category, field='name',
+                                                separator='|')
+            )
+
+            class Meta:
+                model = Book
+
+        book = Book.objects.first()
+        book.categories.all().delete()
+        categories = '|'.join(Category.objects.all().values_list('name', flat=True))
+        dataset_headers = ["id", "categories"]
+        dataset_row = [book.id, categories]
+        dataset = tablib.Dataset(headers=dataset_headers)
+        dataset.append(dataset_row)
+
+        book_resource = _BookM2MResource()
+        book_resource._meta.skip_unchanged = True
+
+        result = book_resource.import_data(dataset, dry_run=False)
+
+        self.assertFalse(result.has_errors())
+        self.assertEqual(len(result.rows), 1)
+
+    def test_m2m_on_existing_entity_changing_relations(self):
+        class _BookM2MResource(resources.ModelResource):
+            categories = fields.Field(
+                attribute='categories',
+                m2m_add=True,
+                widget=widgets.ManyToManyWidget(Category, field='name',
+                                                separator='|')
+            )
+
+            class Meta:
+                model = Book
+
+        book = Book.objects.first()
+        book.categories.set([Category.objects.first()])
+        dataset_headers = ["id", "categories"]
+        dataset_row = [book.id, Category.objects.last().name]
+        dataset = tablib.Dataset(headers=dataset_headers)
+        dataset.append(dataset_row)
+
+        book_resource = _BookM2MResource()
+        book_resource._meta.skip_unchanged = True
+
+        result = book_resource.import_data(dataset, dry_run=False)
+
+        self.assertFalse(result.has_errors())
+        self.assertEqual(len(result.rows), 1)
+
+    def test_m2m_not_present_in_import_file(self):
+        class _BookM2MResource(resources.ModelResource):
+            categories = fields.Field(
+                attribute='categories',
+                m2m_add=True,
+                widget=widgets.ManyToManyWidget(Category, field='name',
+                                                separator='|')
+            )
+
+            class Meta:
+                model = Book
+
+        dataset_headers = ["id", ]
+        dataset_row = [None]
+        dataset = tablib.Dataset(headers=dataset_headers)
+        dataset.append(dataset_row)
+
+        book_resource = _BookM2MResource()
+        book_resource._meta.skip_unchanged = True
```

Comment:
there's a fair amount of duplication in the tests, perhaps the setup logic could be moved to `setUp()`